### PR TITLE
Implement HTMLCanvasElement.toDataURL for WebGL canvas (fixes #19147)

### LINKED ISSUE #19147: webgl: Implement toDataURL()
I'm starting on this, as it's the source of a lot of errors in the webgl conformance tests.  Open questions so far:

- Can I safely assume that the canvas's Width() and Height() match the offscreen_gl_context's FBO width and height?
- What do I need to do as the equivalent of 2d canvas's origin handling?

This is part of #10209

### PR DESCRIPTION
**Pull request by nox**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20400)

### CODE REVIEW

File: components/script/dom/webglrenderingcontext.rs [hunk 1]
```diff
@@ -1157,6 +1157,37 @@ impl WebGLRenderingContext {
...
     }
+
+    // Used by HTMLCanvasElement.toDataURL
+    pub fn get_image_data(&self, mut width: u32, mut height: u32) -> Option<Vec<u8>> {
+        if !self.validate_framebuffer_complete() {
+            return None;
+        }
+
+        if let Some((fb_width, fb_height)) = self.get_current_framebuffer_size() {
+            if width > fb_width as u32 {
```
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: Use `max` and `min`?

File: components/script/dom/webglrenderingcontext.rs [hunk 2]
```diff
@@ -1157,6 +1157,37 @@ impl WebGLRenderingContext {
             }
         }
     }
+
+    // Used by HTMLCanvasElement.toDataURL
+    pub fn get_image_data(&self, mut width: u32, mut height: u32) -> Option<Vec<u8>> {
+        if !self.validate_framebuffer_complete() {
+            return None;
```
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: No error here?

File: components/script/dom/webglrenderingcontext.rs [hunk 3]
```diff
@@ -1157,6 +1157,37 @@ impl WebGLRenderingContext {
...
+
+        if let Some((fb_width, fb_height)) = self.get_current_framebuffer_size() {
+            if width > fb_width as u32 {
+                width = fb_width as u32;
+            }
+            if height > fb_height as u32 {
+                height = fb_height as u32;
+            }
+        } else {
+            self.webgl_error(InvalidOperation);
```
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: Needs spec quote / link.

File: components/script/dom/htmlcanvaselement.rs [hunk 4]
```diff
@@ -346,11 +346,24 @@ impl HTMLCanvasElementMethods for HTMLCanvasElement {
                     Finite::wrap(self.Height() as f64),
                 )?
             }
+            Some(CanvasContext::WebGL(ref context)) => {
+                if let Some(data) = context.get_image_data(self.Width(), self.Height()) {
```
Comment (emilio) [inline components/script/dom/htmlcanvaselement.rs]: Maybe it's cleaner with `match`?

**Review by emilio — COMMENTED** [review]
Looks... reasonable, but I don't know where the WebGL error there should come from, so want links :)

File: components/script/dom/webglrenderingcontext.rs (hunk 1 above)
Comment (emilio) [reply components/script/dom/webglrenderingcontext.rs]: I guess just min :)

File: components/script/dom/webglrenderingcontext.rs (hunk 2 above)
Comment (nox) [reply components/script/dom/webglrenderingcontext.rs]: `validate_framebuffer_complete` stores one already.

**Review by emilio — APPROVED** [review]
r=me assuming `get_image_data` is just moving code around :)

**Comment by nox** [comment]
Yep `get_image_data` is just to avoid creating an `ImageData` for nothing in the 2D path.

@bors-servo r=emilio p=1

**Comment by nox** [comment]
@bors-servo r=emilio

My first commit was apparently wrong and it was just an unrelated drive-by improvement, I removed it.

**Comment by jdm** [comment]
```
  ▶ OK [expected TIMEOUT] /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/log2/log2_001_to_008.html

  ▶ Unexpected subtest result in /_mozilla/webgl/conformance-1.0.3/conformance/ogles/GL/log2/log2_001_to_008.html:
  └ PASS [expected NOTRUN] Overall test
```
This might just be a test that now intermittently times out.

**Comment by nox** [comment]
Yes, most probably. I'll file the issue soon.

**Comment by nox** [comment]
@bors-servo try

Doing a try run to see if it's a more rampant problem or just related to this one test.

**Comment by nox** [comment]
@jdm Wow, what were the odds that we would stress test Homu…

**Pull request merged**